src/Hybrid/transitions.jl

```diff
@@ -0,0 +1,179 @@
+abstract type AbstractTransition end
+
+"""
+    DiscreteTransition{RT, GT, IT⁻, IT⁺, WT} <: AbstractTransition
+
+Type that encodes the discrete transition:
+
+```math
+    post_d(X) = (R(X ∩ G ∩ I⁻) ⊕ W) ∩ I⁺
+```
+where ``I⁻`` and ``I⁺``  are invariants at the source and the target locations
+respectively, the assignment is of the form ``x^+ := Rx + w``, ``w ∈ W``,
+``x^+ ∈ \\mathbb{R}^m`` are the values after the transition,
+``R ∈ \\mathbb{R}^{m\\times n}`` is the assignment map and
+``W ⊆ \\mathbb{R}^m`` is a closed and bounded convex set of non-deterministic inputs.
+
+### Fields
+
+- `R`  -- assignment map
+- `W`  -- non-deterministic inputs
+- `G`  -- guard set of the transition from the source location to the target location
+- `I⁻` -- invariant at the source location
+- `I⁺` -- invariant at the target location
+"""
+struct DiscreteTransition{RT, WT, GT, IT⁻, IT⁺} <: AbstractTransition
+    R::RT
+    W::WT
+    G::GT
+    I⁻::IT⁻
+    I⁺::IT⁺
+end
+
+reset_map(DT::DiscreteTransition) = ConstrainedAffineMap(R, W)
+guard(DT::DiscreteTransition) = DT.G
+source_invariant(DT::DiscreteTransition) = DT.I⁻
+target_invariant(DT::DiscreteTransition) = DT.I⁺
+
+# constructor when the reset map is the identity
+function DiscreteTransition(; guard::GT, source_invariant::IT⁻, target_invariant::IT⁺) where {GT, IT⁻, IT⁺}
+    n = dim(source_invariant)
+    return DiscreteTransition(Universe(n), ZeroSet(n), guard, source_invariant, target_invariant)
+end
+
+apply(f::DiscreteTransition,
+      R::AbstractLazyReachSet,
+      method::AbstractTMDisjointnessMethod) = _apply(f, set(R), method)
+
+# ==========================================
+# Cases where the reset map is the identity
+# ==========================================
+
+# ---------------------------
+# Intersection method: Exact
+# ---------------------------
+
+# all sets are polyhedral
+function _apply(tr::DiscreteTransition{<:Universe, <:ZeroSet, GT, IT⁻, IT⁺},
+                X::AbstractPolyhedron{N},
```

Comment:
It is probably more useful to dispatch on `AbstractPolytope` and then also return an `HPolytope`.

Reply:
Yes, but the set `X` is always bounded.

Reply:
oh, i see. indeed, i missed that argument.